src/Game.hs

```diff
@@ -82,38 +69,43 @@ checkVictory :: Board -> IO ()
 checkVictory board
     | isStartingLayout board = winGame board
     | otherwise              = do
-        clearScreen
-        setCursorPosition 0 0
+        resetScreen
         playBoard board
 
 -- We won!  Throw confetti, and return to the main menu.
 winGame :: Board -> IO ()
 winGame board = do
-    clearScreen
-    setCursorPosition 0 0
-    putStrLn $ show board
+    resetScreen
+    putStrLn (show board)
     putStrLn "You won the game!"
     putStrLn ""
 
+resetScreen :: IO ()
+resetScreen = do
+  clearScreen
+  setCursorPosition 0 0
+
+data TryRead a = Read a | TryAgain | Quit
+
 -- Prompt for and accept a positive number or Q to quit.
 getPositiveNumber :: String -> IO (Maybe Int)
-getPositiveNumber prompt = do
-    queryUser
-    where parseNumber ""             = queryUser
-          parseNumber "Q"            = return Nothing
-          parseNumber response
-              | all isDigit response = ensurePositive $ read response
-              | otherwise            = reportError
-          ensurePositive number
-              | number > 0           = return $ Just number
-              | otherwise            = reportError
-          reportError = do
-              putStrLn "You must enter a positive number."
-              putStrLn ""
-              queryUser
+getPositiveNumber prompt = queryUser
+
+    where parseNumber :: String -> TryRead Int
+          parseNumber "q"            = Quit
+          parseNumber "Q"            = Quit
+          parseNumber response =
+            case Text.Read.readMaybe response of
+              Just n | n > 0 -> Read n
+              _              -> TryAgain
+
           queryUser = do
               putStr prompt
               hFlush stdout
               response <- getLine
-              parseNumber $ map toUpper response
-
+              case parseNumber response of
+                Read n   -> return (Just n)
+                TryAgain -> do
+                  putStrLn "You must enter a positive number.\n"
+                  queryUser
+                Quit     -> return Nothing
```

Comment:
Great pattern!  Thank you for the lightbulbs and for pointing out `Text.Read.readMaybe`.  That removed the awkward `""` check and simplified the positive number check, and it neatly separated the final response from the validation.  I am pushing this change along with very minor adjustments in spacing and order.